### PR TITLE
Comments

### PR DESCRIPTION
**Pull request by Starbugstone**
Comments working as intended

### CODE REVIEW

File: App/Controllers/Admin/Comments.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+     * @throws \Twig_Error_Runtime
+     * @throws \Twig_Error_Syntax
+     */
+    public function moderateComment(int $commentId)
+    {
+        $this->onlyAdmin();
+
+        $comment = $this->commentModel->getCommentById($commentId);
+
+        $userModel = new UserModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Admin/Comments.php:101]: You can inject your model in the contructor

File: App/Controllers/Admin/Home.php [hunk 2]
```diff
@@ -3,24 +3,227 @@
...
+     * @throws \ReflectionException
+     * @throws \Twig_Error_Loader
+     * @throws \Twig_Error_Runtime
+     * @throws \Twig_Error_Syntax
+     */
     public function index()
     {
-        //testing auth
-        if ($this->auth->isAdmin()) {
-            $this->data['userRole'] = 'Admin';
```
Comment (Oipnet) [inline App/Controllers/Admin/Home.php:13]: Use a constant for 'Admin'

File: App/Controllers/Admin/Home.php [hunk 3]
```diff
@@ -3,24 +3,227 @@
...
-            $this->data['userLevel'] = $this->auth->getUserLevel();
-        } elseif ($this->auth->isUser()) {
-            $this->data['userRole'] = 'User';
```
Comment (Oipnet) [inline App/Controllers/Admin/Home.php:16]: use a constant for 'User'

File: App/Controllers/Ajax/AjaxSlugify.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+     * @throws JsonException
+     */
+    public function slugifyString()
+    {
+        //only admins can update the slug
+        $this->onlyAdmin();
+        $this->onlyPost();
+        $slug = $this->request->getData("slugText-update");
+        $result = array();
+        $slugify = new Slugify();
```
Comment (Oipnet) [inline App/Controllers/Ajax/AjaxSlugify.php]: Inject in your constructor

File: App/Controllers/Ajax/Category.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+        if(!$this->slug->isSlugValid($send["categories_slug"]))
+        {
+            $result["success"] = false;
+            $result["errorMessage"] = "Invalid Slug";
+            echo json_encode($result);
+            die();
+        }
+
+        $categoryModel = new CategoryModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Ajax/Category.php]: inject in your constructor

File: App/Controllers/Ajax/Category.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+        if(!$this->isInt($send["idcategories"])){
+            $result["success"] = false;
+            $result["errorMessage"] = "Invalid ID";
+            echo json_encode($result);
+            die();
+        }
+
+
+        $categoryModel = new CategoryModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Ajax/Category.php]: Inject in your constructor

File: App/Controllers/Ajax/Category.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        if(!$this->isInt($send["idcategories"])){
+            $result["success"] = false;
+            $result["errorMessage"] = "Invalid ID";
+            echo json_encode($result);
+            die();
+        }
+
+        $categoryModel = new CategoryModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Ajax/Category.php]: Inject in your constructor

File: App/Controllers/Ajax/Config.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        $this->onlyPost();
+        //preparing our return results
+        $result = array();
+        $result['successId'] = [];
+        $result['errorId'] = [];
+        $result['success'] = true;
+        $configUpdateJson = $this->container->getRequest()->getData('config-update');
+        $configUpdate = json_decode($configUpdateJson);
+
+        $configModel = new ConfigModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Ajax/Config.php]: inject in your constructor

File: App/Controllers/Ajax/ImageUpload.php [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+use Cocur\Slugify\Slugify;
+use Core\AjaxController;
+
+class ImageUpload extends AjaxController
+{
+    /**
+     * @var string the image upload folder, must be writable
+     */
+    private $imageFolder = "uploaded_images/";
```
Comment (Oipnet) [inline App/Controllers/Ajax/ImageUpload.php]: There is a config file ?
You can use it ?

File: App/Controllers/Ajax/PostVerification.php [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        $postSlug = $this->request->getData("postSlug");
+        $postId = (int)$this->request->getData("postId");
+
+        $data = false;
+        if (!$this->slug->isSlugValid($postSlug)) {
+            echo json_encode($data);
+            die();
+        }
+
+        $postModel = new PostModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Ajax/PostVerification.php]: Inject in your controller

File: App/Controllers/Ajax/Tag.php [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        $tagUpdateJson = ($this->request->getData('tag-update'));
+        $tagUpdate = json_decode($tagUpdateJson);
+
+        //Converting our array of objects to simple array
+        $send = array();
+        foreach ($tagUpdate as $item) {
+            $send[$item->name] = $item->value;
+        }
+
+        $tagModel = new TagModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Ajax/Tag.php:34]: Inject in your controller

File: App/Controllers/Ajax/User.php [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+            throw new JsonException("Empty email");
+        }
+
+        try {
+            $return = !$this->userModel->isEmailUsed($email);
+        } catch (\Exception $e) {
+            throw new JsonException($e->getMessage());
+        }
+        echo json_encode($return);
+
```
Comment (Oipnet) [inline App/Controllers/Ajax/User.php:44]: empty line

File: App/Controllers/Author.php [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+     * @throws \ErrorException
+     * @throws \ReflectionException
+     * @throws \Twig_Error_Loader
+     * @throws \Twig_Error_Runtime
+     * @throws \Twig_Error_Syntax
+     */
+    public function posts(int $authorId, string $page = "page-1")
+    {
+        $postModel = new PostModel($this->container);
+        $userModel = new UserModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Author.php]: Inject in constructor

File: App/Controllers/Home.php [hunk 14]
```diff
@@ -12,11 +16,114 @@
...
+     * @throws \Twig_Error_Loader
+     * @throws \Twig_Error_Runtime
+     * @throws \Twig_Error_Syntax
+     */
     public function index()
     {
-        $Includes = new \App\Models\Includes($this->container);
-        $this->data['navigation'] = $Includes->getMenu();
-        $this->data['jumbotron'] = $Includes->getJumbotron();
+        $frontPostModel = new PostModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Home.php]: inject in controller

File: App/Controllers/Login.php [hunk 15]
```diff
@@ -2,28 +2,338 @@
...
+        }
+
+
+        $this->cookie->deleteCookie("rememberMe");
+        $this->session->destroySession();
+        $this->alertBox->setAlert('Disconnected');
+        $this->response->redirect();
+    }
+
+    /*
```
Comment (Oipnet) [inline App/Controllers/Login.php]: Remove temp function

File: App/Controllers/Post.php [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+        $this->data['configs'] = $this->siteConfig->getSiteConfig();
+        $this->data['post'] = $posts;
+        $this->data['postTags'] = $this->tagModel->getTagsOnPost($postId);
+        $this->data['navigation'] = $this->siteConfig->getMenu();
+        $this->data["comments"] = $this->commentModel->getCommentsListOnPost($postId, $pagination["offset"],
+            $linesPerPage);
+        $this->data['pagination'] = $pagination;
+
+        $this->renderView('post');
+
```
Comment (Oipnet) [inline App/Controllers/Post.php:80]: empty line

File: App/Controllers/Tag.php [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+     * @throws \ErrorException
+     * @throws \ReflectionException
+     * @throws \Twig_Error_Loader
+     * @throws \Twig_Error_Runtime
+     * @throws \Twig_Error_Syntax
+     */
+    public function posts(int $tagId, string $page = "page-1")
+    {
+        $postModel = new PostModel($this->container);
+        $tagModel = new TagModel($this->container);
```
Comment (Oipnet) [inline App/Controllers/Tag.php:36]: inject in your controller

File: App/Modules/SendMail.php [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+        // Create the Transport for mail sending
+        //$config = $this->siteConfig->getSiteConfig();
+        $this->transport = (new Swift_SmtpTransport($this->siteConfig["SMTP_server"], (int)$this->siteConfig["SMTP_port"]))
+            ->setUsername($this->siteConfig["SMTP_user"])
+            ->setPassword($this->siteConfig["SMTP_pass"])
+        ;
+
+        // Create the Mailer using your created Transport
+        $this->mailer = new Swift_Mailer($this->transport);
+
```
Comment (Oipnet) [inline App/Modules/SendMail.php]: empty line

File: App/Modules/SendMail.php [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    public function __construct(Container $container)
+    {
+        parent::__construct($container);
+
+        $config = new SiteConfig($this->container);
+        $this->siteConfig = $config->getSiteConfig();
+
+        // Create the Transport for mail sending
+        //$config = $this->siteConfig->getSiteConfig();
```
Comment (Oipnet) [inline App/Modules/SendMail.php:27]: remove unused code

File: App/Modules/SendMail.php [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+     * @param string $subject
+     * @param string $message
+     * @param string|null $from
+     * @return int
+     */
+    public function send(string $to, string $subject, string $message, string $from = null)
+    {
+        // Create a message
+        $message = (new Swift_Message($subject))
+            //->setFrom([$this->siteConfig["SMTP_from"]])
```
Comment (Oipnet) [inline App/Modules/SendMail.php]: unused code

File: App/Modules/Slug.php [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+use Core\Modules\Module;
+
+class Slug extends Module{
+    public function isSlugValid(string $slug):bool
+    {
+        $slugify = new Slugify();
+        $validSlug = $slugify->slugify($slug);
+
+        return $slug === $validSlug;
+
```
Comment (Oipnet) [inline App/Modules/Slug.php]: empty line

Comment (Starbugstone) [reply App/Modules/Slug.php]: Resolved in branch Security_and_bug_fixes

File: App/Modules/SendMail.php (hunk 20 above)
Comment (Starbugstone) [reply App/Modules/SendMail.php]: Resolved in branch Security_and_bug_fixes

File: App/Modules/SendMail.php (hunk 19 above)
Comment (Starbugstone) [reply App/Modules/SendMail.php:27]: Resolved in branch Security_and_bug_fixes

File: App/Modules/SendMail.php (hunk 18 above)
Comment (Starbugstone) [reply App/Modules/SendMail.php]: Resolved in branch Security_and_bug_fixes

File: App/Controllers/Tag.php (hunk 17 above)
Comment (Starbugstone) [reply App/Controllers/Tag.php:36]: Resolved in branch Security_and_bug_fixes

File: App/Controllers/Ajax/Category.php (hunk 5 above)
Comment (Starbugstone) [reply App/Controllers/Ajax/Category.php]: Resolved in branch Security_and_bug_fixes

File: App/Controllers/Ajax/AjaxSlugify.php (hunk 4 above)
Comment (Starbugstone) [reply App/Controllers/Ajax/AjaxSlugify.php]: Resolved in branch Security_and_bug_fixes

File: App/Controllers/Admin/Home.php (hunk 3 above)
Comment (Starbugstone) [reply App/Controllers/Admin/Home.php]: Resolved in branch Security_and_bug_fixes

File: App/Controllers/Admin/Home.php (hunk 2 above)
Comment (Starbugstone) [reply App/Controllers/Admin/Home.php:13]: Resolved in branch Security_and_bug_fixes

File: App/Controllers/Admin/Comments.php (hunk 1 above)
Comment (Starbugstone) [reply App/Controllers/Admin/Comments.php:101]: Resolved in branch Security_and_bug_fixes

**Pull request merged**